File: backend/models/Collection.ts
import { Document, Schema, model } from 'mongoose';

export enum CollectionPermission {
  PUBLIC = 'PUBLIC',
  PRIVATE = 'PRIVATE'
}

export interface ICollection {
  name: string;
  description?: string; // User-provided description
  summary?: string;     // LLM-generated summary of all documents (L2)
  permission: string;
  createdBy: string;
  createdAt?: Date;  // Optional since Mongoose adds it automatically
  updatedAt?: Date;  // Optional since Mongoose adds it automatically
  // add other fields as needed
}

export type CollectionDocument = ICollection & Document;

const collectionSchema = new Schema<CollectionDocument>({
  name: { type: String, required: true },
  description: { type: String },
  summary: { type: String },
  permission: { type: String, required: true },
  createdBy: { type: String, required: true },
  // add default values if needed
}, { timestamps: true });

// Add unique index for collection names
collectionSchema.index({ name: 1 }, { unique: true });

// Add index for better query performance
collectionSchema.index({ createdBy: 1, permission: 1 });

// Add validation for collection name
collectionSchema.pre('save', function(next) {
  if (!this.name || this.name.trim().length === 0) {
    throw new Error('Collection name cannot be empty');
  }
  
  // Sanitize collection name (remove special characters that might cause issues)
  this.name = this.name.trim().replace(/[<>:"/\\|?*]/g, '_');
  
  next();
});

export const CollectionModel = model<CollectionDocument>('Collection', collectionSchema);  

File: backend/models/collection.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from enum import Enum

class CollectionPermission(str, Enum):
    PUBLIC = 'PUBLIC'
    PRIVATE = 'PRIVATE'

class Collection(BaseModel):
    id: Optional[str] = Field(None, alias="_id")
    name: str
    permission: CollectionPermission
    createdBy: str
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True
        json_schema_extra = {
            "example": {
                "name": "My Document Collection",
                "permission": "PRIVATE",
                "createdBy": "user123"
            }
        } 